src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.cs

```diff
@@ -413,13 +413,36 @@ protected void InitializeReferences()
             try
             {
                 runner.InitializeTimeout(internalMatchTimeout);
-                runner.InitializeForScan(this, input, startat, quick: true);
+                runner.InitializeForScan(this, input, startat, quick);
+
+                int stoppos = RightToLeft ? 0 : input.Length;
+
+                // If previous match was empty or failed, advance by one before matching.
+                if (prevlen == 0)
+                {
+                    if (runner.runtextstart == stoppos)
+                    {
+                        return RegularExpressions.Match.Empty;
+                    }
+
+                    runner.runtextpos += RightToLeft ? -1 : 1;
+                }
 
                 runner.Scan(input);
 
                 // If runmatch is null it means that an override of Scan didn't implement it correctly, so we will
                 // let this null ref since there are lots of ways where you can end up in a erroneous state.
-                return runner.runmatch!.FoundMatch ? null : RegularExpressions.Match.Empty;
+                if (runner.runmatch!.FoundMatch)
+                {
+                    if (quick)
+                    {
+                        return null;
+                    }
+                    runner.runmatch.Tidy(runner.runtextpos, 0);
```

Comment:
What would happen if we didn't Tidy? Is that needed even if beginning is 0 and captures aren't used?

Reply:
From the things we need with EnumerateMatches, Tidy sets Index, Length and _textpos. Index and Length should be trivial to get without calling Tidy, since we could just take a look at the internal field _matches and index in the right position to get both values, but _textpos would be more complicated. When calling Tidy, it gets set to runner.runtextpos and from EnumerateMatches we don't have access to the runner so we would need to compute this value, which is probably doable if we had Index and Length but we would also need to take into consideration things like RightToLeft and the like. This is why it felt easier to just call Tidy and get all those three fields computed for us.

We need _textpos because that is what we set _startAt to for our next iteration of the enumerator.

Reply:
ok